### PR TITLE
Add graphql support to node backend

### PR DESCRIPTION
**Pull request by larrylin17**
pass local testing
pass CirculeCI build

### CODE REVIEW

**Comment by bmonkman** [comment]
How will this be templated so that you can choose either GraphQL or REST?

File: templates/src/app/graphql/utils.js [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+};
+
+module.exports.createStore = () => {
+  const Op = SQL.Op;
+  const operatorsAliases = {
+    $in: Op.in,
+  };
+
+  const db = new SQL('database', 'username', 'password', {
+    dialect: 'sqlite',
```
Comment (bmonkman) [inline templates/src/app/graphql/utils.js:38]: There's already support for postgres/mysql in this project, can you use that instead?

**Comment by bmonkman** [comment]
store.sqlite should not be checked in

**Comment by bmonkman** [comment]
Can you strip down the functionality a bit? It doesn't need to be a full implementation of a feature (like you've done with "launch") but just an example of how to implement a query and mutation in GraphQL. It's less important to have a full usable feature and more important to show a good example that is going to be easy for a user to adapt when they create their own API.

**Comment by bmonkman** [comment]
Can you add some documentation? 
- Links / resources for GraphQL
- How is the project structured?
- How does a user define their own schema?
- Which pieces of the code are just examples that a user can delete and which are critical functionality?
- etc.

File: templates/src/app/graphql/utils.js (hunk 1 above)
Comment (larrylin17) [reply templates/src/app/graphql/utils.js:38]: yes, I am working on that.

**Comment by larrylin17** [comment]
> Can you add some documentation?
> 
> * Links / resources for GraphQL
> * How is the project structured?
> * How does a user define their own schema?
> * Which pieces of the code are just examples that a user can delete and which are critical functionality?
> * etc.

Yes, I can add them.

**Comment by larrylin17** [comment]
> store.sqlite should not be checked in

Has removed. Thanks.

**Comment by larrylin17** [comment]
> How will this be templated so that you can choose either GraphQL or REST?

added the following 2 snippets in zero-module.yml. Does it look good?
```  
  - field: graphql
    label: Enable graphql and apollo server?
    default: yes
    options:
      - "yes"
      - "no"
 ```
```
  - action: ignoreFile
    matchField: graphql
    whenValue: "no"
    data:
    - src/app/graphql
```

File: templates/src/app.js [hunk 2]
```diff
@@ -1,31 +1,65 @@
...
+  typeDefs,
+  resolvers,
+  dataSources: () => ({
+    launchAPI: new LaunchAPI(),
+    userAPI: new UserAPI({ store })
+  }),
+});
+server.applyMiddleware({ app });<% end %>
+
+<%if eq (index .Params `fileUploads`) "yes" %>app.use("/file", fileRoutes);<% end %>
```
Comment (bmonkman) [inline templates/src/app.js:46]: Could you add the functionality from the file and auth routes to the gql schema as well? We will want to expose the same functionality regardless of the api type they choose

File: zero-module.yml [hunk 3]
```diff
@@ -90,6 +90,12 @@ parameters:
     options:
       - "yes"
       - "no"
+  - field: graphql
+    label: Enable graphql and apollo server?
```
Comment (bmonkman) [inline zero-module.yml:94]: How about changing this to `apiType`, the label to "What type of API do you want to expose?" and have the options be "graphql" and "rest"?

Comment (larrylin17) [reply zero-module.yml:94]: sounds better. I am changing it.

File: templates/src/app.js (hunk 2 above)
Comment (larrylin17) [reply templates/src/app.js:46]: Sure. I can make it.

File: zero-module.yml (hunk 3 above)
Comment (larrylin17) [reply zero-module.yml]: Solved

File: templates/src/app/graphql/utils.js (hunk 1 above)
Comment (larrylin17) [reply templates/src/app/graphql/utils.js]: Has changed db to postgres / mysql

File: templates/src/app/file/index.js [hunk 4]
```diff
@@ -1,43 +1,37 @@
...
 var aws = require("aws-sdk");
-var cfsign = require("aws-cloudfront-sign");
 
-var router = Router()
+var router = Router();
 var s3 = new aws.S3();
 
-router.get("/presigned/:key", (req, res) => {
+router.get("/presigned", (req, res) => {
+  var bucketName = (req.query.bucket)?(req.query.bucket):process.env.AWS_S3_DEFAULT_BUCKET;
```
Comment (bmonkman) [inline templates/src/app/file/index.js:8]: This is a security risk, we shouldn't let a user choose which bucket to upload to.

File: templates/src/app/file/index.js [hunk 5]
```diff
@@ -1,43 +1,37 @@
...
+    Key: req.query.key,
+    Expires: 60 * 60, // 60 minutes
   };
-
-  var url = cfsign.getSignedUrl(
-    `https://files.${process.env.DOMAIN}/${req.params.key}`,
-    params
-  );
-
+  const url = s3.getSignedUrl('getObject',params);
```
Comment (bmonkman) [inline templates/src/app/file/index.js:29]: Why was this changed to s3 urls instead of cloudfront?

File: templates/src/app/graphql/datasources/AuthAPI.js [hunk 6]
```diff
@@ -0,0 +1,20 @@
+const { RESTDataSource } = require('apollo-datasource-rest');
+
+class AuthAPI extends RESTDataSource {
+    constructor() {
+        super();
+        this.baseURL = 'http://localhost:3000';
```
Comment (bmonkman) [inline templates/src/app/graphql/datasources/AuthAPI.js:6]: Why is this specifying localhost and a static port?

File: templates/src/app.js [hunk 7]
```diff
@@ -1,31 +1,60 @@
...
+app.use(express.urlencoded({ extended: false }));
+
+<%if eq (index .Params `apiType`) "graphql" %>
+const store = createStore();
+
+const server = new ApolloServer({
+  typeDefs,
+  resolvers,
+  dataSources: () => ({
+    launchAPI: new LaunchAPI(),
```
Comment (bmonkman) [inline templates/src/app.js:31]: Can we just get rid of the launch stuff? Seems like it's not really adding much value since the other stuff give some reasonable examples of implementing resolvers.

File: templates/src/app/file/index.js (hunk 4 above)
Comment (larrylin17) [reply templates/src/app/file/index.js:8]: Removed the option. Thanks.

File: templates/src/app/graphql/datasources/AuthAPI.js (hunk 6 above)
Comment (larrylin17) [reply templates/src/app/graphql/datasources/AuthAPI.js:6]: Didn't find the constant variable to the baseURL yesterday. So l leave it there temporarily. I will replace it today. Thanks.

File: templates/src/app.js (hunk 2 above)
Comment (larrylin17) [reply templates/src/app.js:39]: Wrapped /file and /auth APIs with graphql.

File: templates/src/app/graphql/datasources/AuthAPI.js (hunk 6 above)
Comment (larrylin17) [reply templates/src/app/graphql/datasources/AuthAPI.js:6]: add env variable GRAPHQL_BASE_URL to Kubernetes and replaced hared code with env variable.

File: templates/kubernetes/overlays/production/kustomization.yml [hunk 8]
```diff
@@ -19,3 +19,4 @@ configMapGenerator:
     - ENVIRONMENT=production
     - DOMAIN=<% index .Params `productionHostRoot` %>
     - S3_BUCKET=files.<% index .Params `productionHostRoot` %>
+    - GRAPHQL_BASE_URL=https://<% index .Params `productionBackendSubdomain` %><% index .Params `productionHostRoot` %>
```
Comment (bmonkman) [inline templates/kubernetes/overlays/production/kustomization.yml:22]: Please add a template block around these so they are not included if the user chose REST

File: templates/src/app.js (hunk 7 above)
Comment (larrylin17) [reply templates/src/app.js]: Removed LaunchAPI. Thanks.

File: templates/kubernetes/overlays/production/kustomization.yml (hunk 8 above)
Comment (larrylin17) [reply templates/kubernetes/overlays/production/kustomization.yml]: This env variable isn't needed since all APIs wrappers get removed.

File: templates/src/app/file/index.js (hunk 5 above)
Comment (larrylin17) [reply templates/src/app/file/index.js]: The cloudfront in /file in the previous version didn't return presigned url yet. Using s3 is able to get presigned urls. It looks Cloudfront needs more configuration on AWS. So I temporarily use S3.

File: templates/src/middleware/auth/jwtDecoder.js [hunk 9]
```diff
@@ -0,0 +1,20 @@
+const expressJWT = require("express-jwt");
+const pathToRegexp = require('path-to-regexp');
+const { jwtSecret } = require("../../conf");
+
+/*the middleware to verify token and parse token to req.user, 
+please visit https://github.com/auth0/express-jwt for the more details */
+const jwtDecoder = expressJWT({
```
Comment (bmonkman) [inline templates/src/middleware/auth/jwtDecoder.js:7]: Why is there JWT code here? The service will never receive a JWT, as it's handled by the access proxy.

File: templates/src/service/file.js [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+        };
+    }
+
+    getDownloadPresignedUrl(key){
+        const params = {
+            Bucket: process.env.S3_BUCKET,
+            Key: key,
+            Expires: 60 * 60, // 60 minutes
+        };
+        const url = s3.getSignedUrl('getObject',params);
```
Comment (bmonkman) [inline templates/src/service/file.js:27]: This should be using Cloudfront signing, not S3

File: templates/src/graphql/schema.js [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+      url: String
+      method: String 
+    }
+
+    type PresignedUrls {
+      upload: PresignedUrl
+      download: PresignedUrl
+    }
+
+    type Status {
```
Comment (bmonkman) [inline templates/src/graphql/schema.js:20]: Can you please remove this from the schema? We shouldn't be exposing health check endpoints through GraphQL

File: templates/src/graphql/schema.js [hunk 12]
```diff
@@ -0,0 +1,51 @@
+const { gql } = require('apollo-server-express');
+
+const typeDefs = gql`
```
Comment (bmonkman) [inline templates/src/graphql/schema.js:3]: Can you please add some documentation to the schema? https://www.apollographql.com/docs/apollo-server/schema/schema/#documentation-strings

File: templates/src/graphql/resolvers.js [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+        },
+        status: (_, {}, {}) => {
+            var podName = (process.env.POD_NAME)?process.env.POD_NAME:"zero-node-backend";
+            return {
+                ready: "OK",
+                alive: "OK",
+                podName: podName
+            };
+        },
+        bookedTrips: (_, { }, context) => {
```
Comment (bmonkman) [inline templates/src/graphql/resolvers.js:27]: Can you split the "Trip" resolvers out into a separate file so they are easy to remove when the user wants to use it? I'm not convinced we even need this Trip stuff but if we have it, it's only purpose is as an example so it needs to be easy to fully remove without having to change a bunch of different files.

Comment (bmonkman) [reply templates/src/graphql/resolvers.js:27]: If we do keep it, maybe we should tag all the Trip stuff with a certain comment so it's easy for people to search for?

Comment (larrylin17) [reply templates/src/graphql/resolvers.js:19]: split "Trip" from the default schema and resolver.

File: templates/src/graphql/schema.js (hunk 11 above)
Comment (larrylin17) [reply templates/src/graphql/schema.js]: Removed status query from schema and resolver

File: templates/src/middleware/auth/jwtDecoder.js (hunk 9 above)
Comment (larrylin17) [reply templates/src/middleware/auth/jwtDecoder.js:7]: I thought access proxy may pass the access token to the endpoints instead of the plain header.  Otherwise, the endpoints are directly exposed without security protection if the access bypasses the proxy.

File: templates/src/graphql/schema.js (hunk 12 above)
Comment (larrylin17) [reply templates/src/graphql/schema.js]: Added documentation to the schema. Can you please proofread them? Thanks.

File: templates/src/service/file.js (hunk 10 above)
Comment (larrylin17) [reply templates/src/service/file.js:27]: start working on that. Thanks.

File: templates/src/middleware/auth/jwtDecoder.js (hunk 9 above)
Comment (larrylin17) [reply templates/src/middleware/auth/jwtDecoder.js]: restored the authentication component and added the whitelist filter.

File: templates/src/service/file.js (hunk 10 above)
Comment (larrylin17) [reply templates/src/service/file.js]: Replaced s3 to Cloudfront. Thanks.

File: templates/src/app.js [hunk 14]
```diff
@@ -1,31 +1,40 @@
...
 app.use(morgan("combined"));
+app.use(express.json());
+app.use(express.urlencoded({ extended: true }));
+
+<%if eq (index .Params `userAuth`) "yes" %>app.use(authMiddleware);
+app.use("/auth", authRoutes);<% end %>
+
+<%if eq (index .Params `fileUploads`) "yes" %>app.use("/file", fileRoutes);<% end %>
+
+<%if eq (index .Params `userAuth`) "yes" %><% end %>
```
Comment (bmonkman) [inline templates/src/app.js:21]: Nothing in this template block

Comment (larrylin17) [reply templates/src/app.js]: Removed. Thanks

File: templates/src/graphql/default.graphql [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+"""
+Type Query defines all query / read methods that clients can call. 
+These methods must be implemented in resolver as well.
+"""
+type Query {
+  """
+  Get all presigned urls for a specific file 
+  that has been stored or will be stored on a cloud storage
+  """
+  signedUrls(key: String!): SignedUrls
```
Comment (bmonkman) [inline templates/src/graphql/default.graphql:41]: I don't think there is going to be a common use case to get both of these at the same time. Mind splitting it into two queries?

File: templates/src/middleware/auth/index.js [hunk 16]
```diff
@@ -4,22 +4,34 @@ const authMiddleware = (req, res, next) => {
...
+        email: req.headers["x-user-email"],
+      };
+      next();
+    }
   }
+
 };
 
+//whitelist defines the paths that will not pass the authentication middleware
+const auth_whitelist = ["/status/ready", "/status/alive", "/status/about"];
```
Comment (bmonkman) [inline templates/src/middleware/auth/index.js:29]: The other variables are camel case. Also if you're changing it, mind calling it something like `authAllowlist`?

**Comment by bmonkman** [comment]
Just those last two things, I think!

File: templates/src/middleware/auth/index.js (hunk 16 above)
Comment (larrylin17) [reply templates/src/middleware/auth/index.js]: Renamed it to authAllowlist. Thanks.

File: templates/src/graphql/default.graphql (hunk 15 above)
Comment (larrylin17) [reply templates/src/graphql/default.graphql]: split query signedUrls to downloadSignedUrl and uploadSignedUrl. Thanks.

**Review by bmonkman — APPROVED** [review]
